process.c

```diff
@@ -5728,60 +5755,47 @@ check_gid_switch(void)
 VALUE
 rb_getlogin(void)
 {
-#if ( !defined(USE_GETLOGIN_R) && !defined(USE_GETLOGIN) )
-    return Qnil;
-#else
-    char MAYBE_UNUSED(*login) = NULL;
-
-# ifdef USE_GETLOGIN_R
-
-#if defined(__FreeBSD__)
-    typedef int getlogin_r_size_t;
-#else
-    typedef size_t getlogin_r_size_t;
-#endif
-
+# if defined(USE_GETLOGIN_R) && USE_GETLOGIN_R
     long loginsize = GETLOGIN_R_SIZE_INIT;  /* maybe -1 */
 
     if (loginsize < 0)
         loginsize = GETLOGIN_R_SIZE_DEFAULT;
 
     VALUE maybe_result = rb_str_buf_new(loginsize);
-
-    login = RSTRING_PTR(maybe_result);
-    loginsize = rb_str_capacity(maybe_result);
-    rb_str_set_len(maybe_result, loginsize);
+    struct getlogin_r_args args;
+    args.name = RSTRING_PTR(maybe_result);
+    args.namelen = rb_str_capacity(maybe_result);
+    rb_str_set_len(maybe_result, args.namelen);
 
     int gle;
     errno = 0;
-    while ((gle = getlogin_r(login, (getlogin_r_size_t)loginsize)) != 0) {
 
+    while ((gle = IO_WITHOUT_GVL_INT(nogvl_getlogin_r, &args)) != 0) {
         if (gle == ENOTTY || gle == ENXIO || gle == ENOENT) {
             rb_str_resize(maybe_result, 0);
             return Qnil;
         }
 
-        if (gle != ERANGE || loginsize >= GETLOGIN_R_SIZE_LIMIT) {
+        if (gle != ERANGE || args.namelen >= GETLOGIN_R_SIZE_LIMIT) {
             rb_str_resize(maybe_result, 0);
             rb_syserr_fail(gle, "getlogin_r");
         }
 
-        rb_str_modify_expand(maybe_result, loginsize);
-        login = RSTRING_PTR(maybe_result);
-        loginsize = rb_str_capacity(maybe_result);
+        rb_str_modify_expand(maybe_result, args.namelen);
+        args.name = RSTRING_PTR(maybe_result);
+        args.namelen = rb_str_capacity(maybe_result);
```

Comment:
Probably this should be better to cast to `getlogin_r_size_t`.
This is an implicit down-cast, or different signedness.